cli/command/utils.go

```diff
@@ -76,6 +77,67 @@ func PrettyPrint(i any) string {
 
 var ErrPromptTerminated = errdefs.Cancelled(errors.New("prompt terminated"))
 
+type promptOptions struct {
+	disableEcho bool
+}
+
+type PromptOptions func(*promptOptions)
+
+func WithHideUserInput(t bool) PromptOptions {
+	return func(p *promptOptions) {
+		p.disableEcho = t
+	}
+}
+
+// PromptForInput requests input from the user.
+//
+// If the user terminates the CLI with SIGINT or SIGTERM while the prompt is
+// active, the prompt will return an empty string ("") with an ErrPromptTerminated error.
+// When the prompt returns an error, the caller should propagate the error up
+// the stack and close the io.Reader used for the prompt which will prevent the
+// background goroutine from blocking indefinitely.
+func PromptForInput(ctx context.Context, ins *streams.In, outs *streams.Out, message string, opts ...PromptOptions) (string, error) {
+	options := promptOptions{}
+	for _, o := range opts {
+		o(&options)
+	}
+
+	_, _ = fmt.Fprint(outs, message)
+
+	if options.disableEcho {
+		oldState, err := term.SaveState(ins.FD())
+		if err != nil {
+			return "", err
+		}
+		_ = term.DisableEcho(ins.FD(), oldState)
+		defer func() {
+			_ = term.RestoreTerminal(ins.FD(), oldState)
+		}()
+	}
+
+	// On Windows, force the use of the regular OS stdin stream.
+	if runtime.GOOS == "windows" {
+		ins = streams.NewIn(os.Stdin)
+	}
```

Comment:
Why we ignore the passed `ins` on Windows?

Reply:
I'm not too sure tbh - might be a legacy bug which has been resolved. I can maybe check out if we still need to do this fallback on a windows install.

It's possible we don't need it, but I guess keep it for good measure for the time being? :man_shrugging: 
See this comment: https://github.com/docker/cli/blob/master/cli/command/registry.go#L94-L104

Reply:
https://github.com/docker/cli/blob/b83cf582cd3b932b6fce900a54cf0e1e6f61ac17/cli/command/registry.go#L93-L104

Hmm, so the code you linked is still there and already sets the in for the whole `cli` so we don't need to do that again in `PromptForInput`, because `cli.In()` will already be the overriden stdin.

Reply:
Yes, although other places might call `PromptForInput` which might not set this. We also set this in `PromptForConfirmation` so keeping it consistent would be best I think.

Reply:
It completely breaks the assumption that the input will be read from the provided `ins` though.
If the `ins` happens to be something different than stdin, it would still read from stdin on Windows.

This will also make the unit tests from this PR fail on Windows because they provide a `io.Pipe` as an stdin.

Reply:
Alternative could be to pass a bare `io.Reader`, and always handle wrapping it into a `streams.In` internally; perhaps we don't even need the `streams` considering that we already use `moby/term` (which does most / all of the heavy-lifting in the `streams` package.

I'm curious though; if we need the special handling here; did we do something wrong with setting up the CLI and/or elsewhere; we should definitely look at that, and check if this is still needed.

I did a quick search; this code was originally introduced in the Moby repo (before it was moved to docker/cli); in this commit; https://github.com/moby/moby/commit/9c765040a5638429384b56a1b77139c6d4049c75

That's this PR, and related tickets;

- https://github.com/moby/moby/pull/17738
- https://github.com/moby/moby/issues/14336
- https://github.com/moby/moby/issues/14210

That PR was for Windows RS4, so ... really old, and if it was a Windows bug, that side may at least be obsolete (🤞) but we should check!

Reply:
I've removed the windows check in the last commit https://github.com/docker/cli/pull/5168/commits/668f8743225a610f07411f80258ba3653f477810 - but still need to test a build on windows. I guess I'll need to install some older (supported) versions to ensure the majority of users won't be impacted.

Reply:
I've tested on a recent version of windows 10 ([19045.4529](https://support.microsoft.com/en-us/topic/june-11-2024-kb5039211-os-builds-19044-4529-and-19045-4529-f7e528c9-5e9f-4cd8-9161-704708448517)) and seems to work fine. I removed these lines https://github.com/docker/cli/blob/b83cf582cd3b932b6fce900a54cf0e1e6f61ac17/cli/command/registry.go#L93-L104 locally and tested the login flow. Works fine in powershell and command prompt